File: src/strategies/funcs.ts
import { Strategy } from "@/classes/strategy";
import { data } from "@/data/data";
import {
    MAKER_FEE_RATE,
    TAKER_FEE_RATE,
    slPercent,
    P_DIFF,
    minDiff,
    isMarket,
    cancelOnCond,
} from "@/utils/constants";
import { calcEntryPrice } from "@/utils/funcs2";
import {
    getCoinPrecision,
    getPricePrecision,
    toFixed,
} from "@/utils/functions";
import { IObj } from "@/utils/interfaces";

let enterTs = "";
interface IBalProfitCalc {
    balance: number;
    base: number;
    exit: number;
    data: IObj;
    ccy: string;
    row: IObj;
    prevCandle: IObj;
    prevRow: IObj;
    maker: number;
    pricePrecision: number;
}
export const balProfitCalc: (props: IBalProfitCalc) => any[] = ({
    balance,
    base,
    exit,
    data,
    row,
    maker,
    pricePrecision,
    prevCandle,
    prevRow,
}) => {
    /**
     * We are selling
     * Returns balance, profit, data
     *  */
    console.log("LOG");
    console.log(base, exit);
    let newBalance = toFixed(base * exit, pricePrecision);
    //newBalance -= newBalance * maker;
    //newBalance = toFixed(newBalance, pricePrecision);
    const fee = newBalance * maker;
    let profit = newBalance - fee - balance;
    let profitPercentage = (profit / balance) * 100; //(exit - entry) / entry * 100
    profitPercentage = Number(profitPercentage.toFixed(4));
    balance = newBalance - fee; //+= profit;
    profit = toFixed(profit, pricePrecision);

    data["data"][row["ts"]] = {
        side: `sell \t {h:${prevCandle.h} -> ${prevRow.h}, l: ${prevCandle.l} -> ${prevRow.l}}`,
        enterTs,
        fill: exit,
        c: exit,
        balance: `${newBalance} \t -${fee}`,
        profit: `${profit}\t${profitPercentage}%`,
    };
    //for (let k of Object.keys(data.data))
    return [balance, profit, data];
};
export const strategy = ({
    df,
    candles,
    balance,
    buyCond,
    sellCond,
    lev = 1,
    pair,
    maker = MAKER_FEE_RATE,
    taker = TAKER_FEE_RATE,
}: {
    df: IObj[];
    candles: IObj[];
    balance: number;
    buyCond: (row: IObj, df?: IObj[], i?: number) => boolean;
    sellCond: (row: IObj, entry: number, df?: IObj[], i?: number) => boolean;
    pair: string[];
    maker: number;
    taker: number;
    lev?: number;
}) => {
    let pos = false;
    let cnt = 0,
        gain = 0,
        loss = 0;

    let mData: IObj = { data: {} },
        _data: IObj;
    console.log("CE_SMA: BEGIN BACKTESTING...\n");
    let entry: number = 0,
        entryLimit: number | null = null,
        exitLimit: number | null = null,
        base: number = 0,
        sl: number = 0,
        exit: number = 0,
        profit: number = 0;

    const pricePrecision = getPricePrecision(pair, "bybit");
    const basePrecision = getCoinPrecision(pair, "sell", "bybit");
    balance = toFixed(balance, pricePrecision);
    df = df.slice(20);
    candles = candles.slice(20);

    for (let i = 1; i < df.length; i++) {
        const prevRow = df[i - 1],
            prevCandle = candles[i - 1],
            row = candles[i];
        console.log(`\n${prevRow.ts}\n`);
        if (entryLimit && !pos) {
            console.log(
                `[${prevCandle.ts} \t H: ${prevCandle.h} \t E: ${entryLimit} \t L: ${prevCandle.l}\n]`
            );

            const cond = prevCandle.l <= entryLimit; // && entryLimit < prevCandle.h
            if (cond) {
                /* Fill buy order */
                entry = entryLimit;
                base = (balance / entry) * (1 - taker);
                base = toFixed(base, basePrecision);

                mData["data"][prevCandle["ts"]] = {
                    side: `buy \t {h:${prevCandle.h}, l: ${prevCandle.l}}`,
                    fill: entryLimit,
                    enterTs,
                    c: entry,
                    balance: `[${balance}] \t ${base} \t -${base * taker!}`,
                };

                pos = true;
            } else if (sellCond(prevRow, entry)){
                /* Cancel buy order If price goes down 2% */
                console.log(`${row.ts} \t CANCELLING BUY ORDER`);
                mData["data"][prevCandle.ts] = {
                    side: `buy \t {h:${prevCandle.h}, l: ${prevCandle.l}}`,
                    c: entryLimit,
                    pricePrecision,
                    cancel: true,
                };
                entryLimit = null;
            }
        } else if (exitLimit) {
            const cond =
                /* prevCandle.l < exitLimit && exitLimit < prevCandle.h || */ prevCandle.h >=
                exitLimit;
                const isLoss = prevCandle.l <= sl

            if (cond || isLoss) {
                /* Fill sell order */
                exit = isLoss && !cond ? sl : exitLimit;
                balance = base * exit * (1 - maker);
                balance = toFixed(balance, pricePrecision);

                mData["data"][prevCandle["ts"]] = {
                    side: `sell \t {h:${prevCandle.h}, l: ${prevCandle.l}}`,
                    fill: exitLimit,
                    enterTs,
                    c: exit,
                    balance: `[${balance}] \t ${base}`,
                };
                /* Position now filled */
                entryLimit = 0;
                base = 0;
                exitLimit = 0;
                pos = false;

                if (exit > entry) gain += 1;
                else loss += 1;
                cnt += 1;
            } else {
                console.log(`${row.ts} \t CANCELLING SELL ORDER`);
                mData["data"][prevCandle.ts] = {
                    side: `sell \t {h:${prevCandle.h}, l: ${prevCandle.l}}`,
                    c: exitLimit,
                    cancel: true,
                };

                exitLimit = null;
            }
        }

        if (!entryLimit && buyCond(prevRow)) {
            /* Place buy order */
            entryLimit = calcEntryPrice(prevCandle, "buy");
            entryLimit = toFixed(entryLimit ?? 0, pricePrecision);
            enterTs = row.ts;
            console.log(`[ ${row.ts} ] \t Limit buy order at ${entryLimit}`);
        } else if (
            !exitLimit &&
            entryLimit &&
            pos /* &&
            sellCond(prevRow, entry) */
        ) {
            /* Place sell order */
            exitLimit = entry * (1 + 5/100) //prevCandle.c * (1 + 3 / 100);
            exitLimit = toFixed(exitLimit, pricePrecision);
            sl = entry * (1 - .5/100) //prevCandle.c * (1 + 3 / 100);
            sl = toFixed(sl, pricePrecision);
            enterTs = row.ts;
            console.log(`[ ${row.ts} ] \t Limit sell order at ${exitLimit}`);
        }
    }

    gain = Number((gain / cnt * 100).toFixed(2))
    loss = Number((loss / cnt * 100).toFixed(2))
    _data = { ...mData, balance, trades: cnt, gain, loss };
    return _data;
};

File: strategies/funcs.py
from data.data import data
from strategies.utils.funcs import fill_buy_order, fill_sell_order
from utils.constants import MAKER_FEE_RATE, SL, TAKER_FEE_RATE, TP, is_market
import pandas as pd

from utils.funcs3 import get_coin_precision, get_price_precision

def strategy(df: pd.DataFrame, balance: float, buy_cond, sell_cond, pair: list[str], maker: float, taker: float):

    pos = False; 
    cnt = 0
    _cnt = 0
    gain = 0
    loss = 0

    m_data = { 'data': [] }
    _data = {}
    entry: float = 0
    entry_limit: float | None = None
    exit_limit: float | None = None
    tp: float | None = None
    base: float = 0
    sl: float | None = None
    exit: float = 0
    enter_ts = ""
    skip = ""
    profit: float = 0

    price_precision = get_price_precision(pair, "bybit")
    base_precision = get_coin_precision(pair, "sell", "bybit")
    print("CE_SMA: BEGIN BACKTESTING...\n")

    for i, row in df.iterrows():
        if i <= 0: continue

        prev_row = df.loc[i - 1]
        print(f"\nTS: {row['ts']}")
        
        def _fill_sell_order(ret):
            nonlocal exit_limit, exit, prev_row, entry, base, balance, price_precision, enter_ts, gain, loss, cnt, m_data, pos, sl, tp, entry_limit
            pos = ret["pos"]
            m_data = ret["m_data"]
            sl = ret["sl"]
            balance = ret["balance"]
            tp = ret["tp"]
            entry_limit = ret["entry_limit"]
            exit_limit = ret.get("exit_limit")
            cnt = ret["cnt"]
            gain = ret["gain"]
            base = ret["base"]
            loss = ret["loss"]
            print({'tp': ret['tp'], 'sl': ret['sl']})
            
        def _fill_buy_order(ret: dict):
            nonlocal m_data, base, balance, _cnt, pos, base, enter_ts, sl, tp
            m_data = ret['m_data']
            base = ret['base']
            balance = ret['balance']
            _cnt = ret['_cnt']
            pos = ret['pos']
            base = ret['base']

            enter_ts = row["ts"]
            tp = round(entry * (1 + TP / 100), price_precision)
            sl = round(entry * (1 - SL / 100), price_precision)

        if skip == prev_row['ts']:
            pass
        
        if not pos and entry_limit:
            if prev_row['l'] <= entry_limit:
               
                entry = entry_limit
                ret = fill_buy_order( entry=entry, prev_row=prev_row, 
                entry_limit = entry_limit, enter_ts=enter_ts, 
                taker = taker, base = base, balance = balance, 
                base_precision=base_precision, m_data = m_data, pos = pos,)
                
                _fill_buy_order(ret)
                
        if pos and (sl or tp):
            print('HAS POS OR SL')

            _tp = tp
            _sl = sl
            _pos = pos

            is_sl = sl and row['l'] <= sl

            if sl and sl < entry and is_sl:
                print("FILL AT SL")
                exit = round(sl, price_precision)

                ret = fill_sell_order(exit_limit, exit, row, entry=entry, base = base, balance = balance, price_precision = price_precision, enter_ts = enter_ts, gain = gain, loss = loss, cnt = cnt, m_data = m_data, pos = pos, sl = sl, tp = tp, entry_limit = entry_limit)
                _fill_sell_order(ret)
            elif tp and tp <= row['h']:
                print("FILL AT TP")
                exit = round(tp, price_precision)

                ret = fill_sell_order(exit_limit, exit, row, entry=entry, base = base, balance = balance, price_precision = price_precision, enter_ts = enter_ts, gain = gain, loss = loss, cnt = cnt, m_data = m_data, pos = pos, sl = sl, tp = tp, entry_limit = entry_limit)
                _fill_sell_order(ret)

            if not pos:
                print({"entry": entry})
                skip = row['ts']

        if (not pos and buy_cond(prev_row)):
            # PLACE MARKET BUY ORDER */
            entry_limit = row['o'] if is_market else prev_row['c']
            entry_limit = round(entry_limit, price_precision)
            enter_ts = row['ts']
            print(f"[ {row.ts} ] {'Market' if is_market else 'Limit'} buy order at {entry_limit}")

            if is_market:
                entry = entry_limit
                ret = fill_buy_order( entry=entry, prev_row=prev_row, 
                entry_limit = entry_limit, enter_ts=enter_ts, 
                taker = taker, base = base, balance = balance, 
                base_precision=base_precision, m_data = m_data, pos = pos,)
                
                _fill_buy_order(ret)

    print(f"TOTAL TRADES: {cnt}")
    cnt = cnt if cnt > 0 else 1
    gain = round(gain * 100 / cnt, 2)
    loss = round(loss * 100 / cnt, 2)
    m_data = {**m_data, 'balance': round(balance , price_precision),
              'trades': cnt, "gain": gain, 'loss': loss}
    return m_data

